composer/profiler/performance_analysis_trace_handler.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Outputs profiling data in JSON trace format."""
+
+import warnings
+from typing import Any, Dict, List, Tuple, Union
+
+import numpy as np
+
+from composer.core.state import State
+from composer.core.time import Timestamp
+from composer.loggers import Logger
+from composer.profiler.profiler_action import ProfilerAction
+from composer.profiler.trace_handler import TraceHandler
+
+__all__ = ['PerformanceAnalyzerTraceHandler']
+
+MINIMUM_DATAPOINTS_NEEDED = 10
+
+
+class PerformanceAnalyzerTraceHandler(TraceHandler):
+    """Meaures if trainer run is model or dataloader bottlenecked."""
+
+    def __init__(self):
+        # Stores wait times for the current epoch
+        self.dataloader_wait_times: Dict[int, Any] = {}
+        self.batch_wait_times: Dict[int, Any] = {}
+        self.last_start_time = None
+        # Dataloader is bottlenecked if yield takes more than 1% of batch compute time and at least
```

Comment:
Should the batch yield time depend on the size of the batch? From the pytorch DL source code, I don't think it does -- it just moves pointers around and increments / decrements counters: https://github.com/pytorch/pytorch/blob/1cd1f48d9311561920c841d322a98b805e49dcc7/torch/utils/data/dataloader.py#L676

Thinking then would the constant overhead check be sufficient? Curious if you found in practice it does depend on the batch size?

I do think it would still be super helpful to report how long (as a %) we spent with dataloader yielding, but on second thought, I'm not sure whether that percentage should be used to justify whether we are DL bottlenecked. CC @bandish-shah?

Reply:
My understanding (haven't traced through the PyTorch code though) is if a dataset is written in a way where decoding happens on fetching an item (eg for images), the yield time becomes a fn of batch size

Reply:
In general, we should assume that we are dataloader bottlenecked any time the model is idle because we are waiting for dataloader to return a batch.  The dataloader latency should include any Python control overhead, which I believe can be dataloader implementation specific.  At some point it would be good to provide a breakdown indicating what exactly the dataloader is spending time on, we can include control overhead as a part of this break down.   

Might be good to chat about this offline but personally I'm not a huge fan of static thresholds.  If the implementation requires them, we should make them configurable by the user because it's not clear if we can come up with good defaults that work in all cases.